sklearn/utils/estimator_checks.py

```diff
@@ -3138,6 +3138,78 @@ def check_requires_y_none(name, estimator_orig, strict_mode=True):
             warnings.warn(warning_msg, FutureWarning)
 
 
+def check_dataframe_column_names_consistency(name, estimator_orig):
+    try:
+        import pandas as pd
+    except ImportError:
+        raise SkipTest("pandas is not installed: not checking "
+                       "column name consistency for pandas")
+
+    def _construct_dataframe(X, columns):
+        return pd.DataFrame(X, columns=columns)
+    _check_column_name_consistency(name, estimator_orig, _construct_dataframe,
+                                   "dataframe")
+
+
+def check_dataarray_column_names_consistency(name, estimator_orig):
+    try:
+        import xarray as xr
+    except ImportError:
+        raise SkipTest("xarray is not installed: not checking "
+                       "column name consistency for xarray")
+
+    def _construct_xarray(X, columns):
+        return xr.DataArray(X, dims=('index', 'columns'),
+                            coords={'columns': columns})
+    _check_column_name_consistency(name, estimator_orig, _construct_xarray,
+                                   "xarray")
+
+
+def _check_column_name_consistency(name, estimator_orig, construct_X,
+                                   array_name):
+    estimator = clone(estimator_orig)
+
+    X_orig, _ = make_regression(random_state=0, n_features=10)
+    X_orig = _enforce_estimator_tags_x(estimator, X_orig)
+    X_orig = _pairwise_estimator_convert_X(X_orig, estimator)
+
+    n_samples, n_features = X_orig.shape
+    names = np.array([f"col_{i}" for i in range(n_features)])
+    X = construct_X(X_orig, names)
+
+    rng = np.random.RandomState(0)
+    if is_regressor(estimator):
+        y = rng.normal(size=n_samples)
+    else:
+        y = rng.randint(low=0, high=2, size=n_samples)
+    y = _enforce_estimator_tags_y(estimator, y)
+
+    estimator.fit(X, y)
+
+    if not hasattr(estimator, "feature_names_in_"):
+        raise ValueError("Estimator does not have a feature_names_in_ "
+                         f"attribute after fitting with a {array_name}")
+
+    if estimator.feature_names_in_ is None:
+        # no names to check
+        return
```

Comment:
I am not sure when is this expected to happen.

At the moment, if I comment out those lines, only `DummyClassifier` / `DummyRegressor` fail. But they have the `"no_validation"` estimator tag. We might want to skip the end of this check only if this tag is set instead.

WDYT?

Reply:
That makes sense. I updated this PR to reflect this change.